src/box/lua/net_box.c

```diff
@@ -123,6 +123,7 @@ enum netbox_state {
 	NETBOX_ERROR             = 4,
 	NETBOX_ERROR_RECONNECT   = 5,
 	NETBOX_CLOSED            = 6,
```

Comment:
This is an intended behavior. The idea is to keep the server code as simple as possible. Making the server hang around until all pending requests have been completed would significantly complicate the code.

I didn't ask product managers (only discussed the issue with @alyapunov), because I don't believe it's PM's job to handle such low-level technical details. For them the visible behavior would be the same in the most common case: when a server is restarted while there are active clients.

Regarding your example, the client could die before executing the request, or after executing the request, but before flushing the send buffer, or after flushing the send buffer, but before the kernel actually sent the request to the peer. In any of those cases the request wouldn't be processed. Besides, the server could die by timeout or be forcefully killed by systemd with SIGKILL. In other words, there can't possibly be a guarantee that all requests have been completed - the feature just provides a best effort attempt to let clients choose another host when the current one is about to be shut down.